students/kristianjf/lesson09/mailroom_5.py

```diff
@@ -0,0 +1,221 @@
+        ds_names = [name for name, donation in self.DATA_STRUCTURE]
+
+        if full_name in ds_names:
+            full_name_index = ds_names.index(full_name)
+            print(f'Found {full_name} in data_structure')
+            while not donation_amount:
+                donation_amount = int(input('Please enter a donation amount: '))
+            self.DATA_STRUCTURE[full_name_index][1].append(donation_amount)
+            print(self.output_string().format(name=full_name, donation=donation_amount))
+        else:
+            while not donation_amount:
+                donation_amount = int(input('Please enter a donation amount: '))
+            self.DATA_STRUCTURE.append((full_name, [donation_amount]))
+            print(self.output_string().format(name=full_name, donation=donation_amount))
+        return self.output_string().format(name=full_name, donation=donation_amount)
+
+    def list_donors(self):
+        '''
+        If the user types ‘list’, show them a list of the donor names and re-prompt
+        '''
+        for entry in self.DATA_STRUCTURE:
+            print(entry[0])
+        return [entry[0] for entry in self.DATA_STRUCTURE]
+
+    def create_a_report(self):
+        '''
+        Creating a Report
+        If the user (you) selected “Create a Report”, print a list of your donors, \
+        sorted by total historical donation amount.
+        '''
+        print('Menu: Create a Report')
+        top_row = ('Donor Name', 'Total Given', 'Num Gifts', 'Average Gift')
+        format_header = '{:<30} | {:<15} | {:<10} | {:<15}\n'
+        format_data = '{:<30} $ {:<15.2f}   {:<10} $ {:<15.2f}\n'
+        format_string = format_header.format(*top_row)
+        for i in sorted(self.DATA_STRUCTURE, key=self.key_index, reverse=True):
+            name, total_given, num_gifts, avg_gift = i[0], sum(i[1]), len(i[1]), sum(i[1]) / len(i[1])
+            format_string += format_data.format(name, total_given, num_gifts, avg_gift)
+        print(format_string)
+        return format_string
+
+    def key_index(self, item):
+        '''Sort by total sum of donation amount'''
+        return sum(item[1])
+
+    def send_letters(self):
+        '''
+        Try to use a dict and the .format() method to do the letter as one big \
+        template rather than building up a big string in parts.
+        In this version, add a function (and a menu item to invoke it), that goes \
+        through all the donors in your donor data structure, generates a thank you \
+        letter, and writes it to disk as a text file.
+        '''
+        for entry in self.DATA_STRUCTURE_DICTS:
+            filename = '_'.join(entry.get('name').split())+'.txt'
+            filenames = []
+            with open(filename, 'w') as outfile:
+                outfile.write(self.output_string().format(**entry))
+                print(f'Writing: {filename}')
+                filenames.append(filename)
+        return filenames
+
+    def output_string(self):
```

Comment:
and this should be in the Donor class -- a Donor knows how to write a letter to its donor.